### ISSUE TITLE
Split aggregate nicknames into a separate file from the omni_config

### ISSUE THREAD

**Issue by seledwards** [GENI-NSF/geni-tools#352]
Instead of putting the default set of aggregate nicknames in the `omni_config`, we should instead put them in another `.ini` file (let's call it `agg_nick_config`).  `agg_nick_config` would be posted on a public website so that it could be download at will.

I believe `ini` files can be cascaded so you would read the `omni_config` and also `agg_nick_config` and you would end up with the same information that you get from the `omni_config` today.  (Further, I suspect we could pretty easily allow people to continue to put custom agg nicknames in their `omni_config` while still getting the default nicknames from the new `agg_nick_config`.)

The advantages of this are:
 * it allows us to update the list of agg nicknames between gcf releases
 * it's quick and easy to implement
 * we don't lose functionality and we don't have to wait for the new CH
 * we don't have to keep the agg nickname info up-to-date in multiple places in the gcf distribution

The question is how do we go about downloading the updated `agg_nick_config` file?
 * The experimenter could manually download it using wget
 * omni-configure.py could be responsible for downloading it the first time.  But subsequent downloads would them require the experimenter to take the time to run a command.
 * `omni.p`y could treat `agg_nick_config` like it treats the GetVersion cache.  For example, if the timestamp on the `agg_nick_config` file is older than some threshold, it could just try to go download the latest version of the file. If the download fails, it would just use the existing file.

Folks seem to like option 3, which is to treat the aggregate nickname file like the GetVersion cache.

Some requirements:
 * gcf should be distributed with a version of the `agg_nick_config`
 * omni-configure.py should update `agg_nick_config` by default when it is run
 * omni.py should update `agg_nick_config` periodically
 * omni.py should allow the user to update `agg_nick_config` on demand
 * omni.py should allow the user to prevent the update of `agg_nick_config` on demand 
 * Don't error or fail if can't get the updated file.  Instead print an INFO level message
 * Users should be able to continue to specify their own aggregate nicknames

Also, the definitive `agg_nick_config` should live at: http://trac.gpolab.bbn.com/gcf/raw-attachment/wiki/Omni/agg_nick_cache

Let's use the GetVersion cache options as a starting point.  The GetVersion cache utilizes the following options.

```
  --NoGetVersionCache   Disable using cached GetVersion results (forces
                        refresh of cache)
  --ForceUseGetVersionCache
                        Require using the GetVersion cache if possible
                        (default false)
  --GetVersionCacheAge=GETVERSIONCACHEAGE
                        Age in days of GetVersion cache info before refreshing
                        (default is 7)
  --GetVersionCacheName=GETVERSIONCACHENAME
                        File where GetVersion info will be cached, default is
                        ~/.gcf/get_version_cache.json
```

 * Specify where `agg_nick_config` is stored locally
   * Set by `--AggNickCacheName`
   * Default: `.gcf/agg_nick_config`
 * To force update of `agg_nick_config`
   * Use `--NoAggNickCache` (if `True` then `--ForceUseAggNickCache` is `False`)
   * Default: `False`
 * To prevent the update of `agg_nick_config`?
   * Use `--ForceUseAggNickCache` (if `True` then  `--NoAggNickCache` is `False`)
   * Default: `False`
 * To set the age of `agg_nick_config`
   * Use `--AggNickCacheAge`
   * Default: 7 days (to match the GetVersion cache age)
 * ADDED: To set the url to query for the `agg_nick_cache`
   * Use `--AggNickDefinitiveLocation`
   * Default: `http://trac.gpolab.bbn.com/gcf/raw-attachment/wiki/Omni/agg_nick_cache`

## Pseudo-Code
```
def update_agg_nick_config():
 """Try to download the definitive version of `agg_nick_config` and store in the specified place."""
 try:
   wget `agg_nick_config`
   cp `agg_nick_config` AggNickCacheName
 except:
   log "INFO: Tried to download latest `agg_nick_config` but could not."
```

`omni.py`
```
# if the user specified to update the cache or if the cache is old,
# then update the cache
if ( NoAggNickCache ) OR ( age(agg_nick_config) <= (now - AggNickCacheAge) ):
  update_agg_nick_config()
```

```  
# load agg_nick_config then load omni_config
load_agg_nick_config() # includes default aggregate nicknames
load_config() # includes custom aggregate nicknames
```

`omni-configure.py`

SEE UPDATE: I don't think this is necessary

```
# always update the cache
update_agg_nick_config()
```

 _Imported from trac ticket [#352](http://trac.gpolab.bbn.com/gcf/ticket/352),  created by sedwards on 08-26-2013 at 17:22, last modified: 05-18-2015 at 10:15_
   CCing: gpo-expt-support@...

**Comment by ahelsing** [comment]
Nice to haves for field level debugging:
 1. At debug level, log which file produced which nickname
 2. Log if nickname cache was updated
 3. Log if it loaded an aggregate nickname cache file or couldn't find it


_Trac comment by ahelsing on 08-29-2013 at 10:52_

**Comment by seledwards** [comment]
I think I already have 2 and 3 (but I'll double check).  And I'll add 1.

Thanks for the suggestion.

_Trac comment by sedwards (github user: seledwards) on 08-29-2013 at 11:40_

**Comment by seledwards** [comment]
Niky made the following good suggestions:

 * we should make sure that the omni_config.sample and the omni_config that is autogenerated by omni_configure still have a section for AM Nicknames so that users know where to add their personal ones
 * When there is a conflict we might want to print a warn message to the user about the conflict and what was actually used
 * Although 7 days seems reasonable, it might be too long for when we add a new rack since for frequent users this means that most probably they will have to manually force the upload when they want to use newly available racks, once. If is it not a very heavy weigh process, once per day might be also reasonable
 * Add some text to the error message when an AM nickname is not found, that will indicate that they might need to update their file.

_Trac comment by sedwards (github user: seledwards) on 08-29-2013 at 15:26_

**Comment by seledwards** [comment]
Replying to [sedwards](comment:6):
> Niky made the following good suggestions:
> 
>  * we should make sure that the omni_config.sample and the omni_config that is autogenerated by omni_configure still have a section for AM Nicknames so that users know where to add their personal ones

DONE

>  * When there is a conflict we might want to print a warn message to the user about the conflict and what was actually used

DONE

>  * Although 7 days seems reasonable, it might be too long for when we add a new rack since for frequent users this means that most probably they will have to manually force the upload when they want to use newly available racks, once. If is it not a very heavy weigh process, once per day might be also reasonable

Changed to ONCE A DAY

>  * Add some text to the error message when an AM nickname is not found, that will indicate that they might need to update their file.

DONE.

_Trac comment by sedwards (github user: seledwards) on 08-30-2013 at 21:45_

**Comment by seledwards** [comment]
Replying to [ahelsing](comment:3):
> Nice to haves for field level debugging:
>  1. At debug level, log which file produced which nickname
>  2. Log if nickname cache was updated
>  3. Log if it loaded an aggregate nickname cache file or couldn't find it

All three are DONE

_Trac comment by sedwards (github user: seledwards) on 08-30-2013 at 21:46_

**Comment by seledwards** [comment]
I believe I've done everything in this ticket except modifying `omni-configure.py` to use `update_agg_nick_config()`.  I think this is unnecessary because `omni.py` will update the file the first time it is run if it doesn't exist already.

Also I added a `--AggNickDefinitiveLocation` to allow you to specify the url to query for the `agg_nick_cache` for.

I added notes in the ticket description to reflect these two modifications.

_Trac comment by sedwards (github user: seledwards) on 08-30-2013 at 21:51_
